src/frontend/lib/storage/provider.ts

```diff
@@ -0,0 +1,204 @@
+                    error: errorData,
+                })
+                throw new Error(errorData.details || 'Failed to get upload URL')
+            }
+
+            const data = await response.json()
+            console.log('Received presigned URL data:', data, {
+                key: data.key,
+                expiresIn: data.expiresIn,
+            })
+
+            return data
+        } catch (error) {
+            console.error('Error getting presigned URL:', error)
+            throw error
+        }
+    }
+
+    private async uploadWithProgress(
+        url: string,
+        file: File,
+        onProgress?: (progress: UploadProgress) => void,
+    ): Promise<Response> {
+        return new Promise((resolve, reject) => {
+            const xhr = new XMLHttpRequest()
+
+            xhr.upload.addEventListener('progress', event => {
+                if (event.lengthComputable && onProgress) {
+                    onProgress({
+                        loaded: event.loaded,
+                        total: event.total,
+                        percentage: (event.loaded / event.total) * 100,
+                    })
+                }
+            })
+
+            xhr.addEventListener('load', () => {
+                const isValidStatus = xhr.status >= 200 && xhr.status < 300
+
+                if (isValidStatus) {
+                    resolve(
+                        new Response(xhr.response, {
+                            status: xhr.status,
+                            headers:
+                                this.config.provider !== Provider.Azure
+                                    ? new Headers({
+                                          ETag:
+                                              xhr.getResponseHeader('ETag') ||
+                                              '',
+                                      })
+                                    : undefined,
+                        }),
+                    )
+                } else {
+                    reject(
+                        new Error(
+                            `Upload failed with status ${xhr.status}: ${xhr.responseText}`,
+                        ),
+                    )
+                }
+            })
+
+            xhr.addEventListener('error', () =>
+                reject(new Error('Network error during upload')),
+            )
+
+            xhr.open('PUT', url)
+            if (this.config.provider === Provider.Azure)
+                xhr.setRequestHeader('x-ms-blob-type', 'BlockBlob')
+            xhr.send(file)
+        })
+    }
+
+    validateConfig(): boolean {
+        const required = ['tokenEndpoint'] as const
+        const missing = required.filter(key => !this.config[key])
+
+        if (missing.length > 0)
+            throw new Error(
+                `Missing required configuration: ${missing.join(', ')}`,
+            )
+
+        return true
+    }
+
+    private handleError(error: unknown): never {
```

Comment:
Issue: The method relies on checking substrings in err.message to determine the error type. This approach is fragile and can break if error messages change or are localized.

Recommendation: Use more structured error handling by checking specific error properties like name, code, or custom properties returned from the backend. For example, if the backend returns an error object with a code property, you can use that to determine the error type:


private handleError(error: unknown): never {
    if (error instanceof UploadError) throw error;

    const err = error as any;
    if (err.code === 'UNAUTHORIZED') {
        throw new UploadError(
            'Unauthorized access to Provider',
            UploadErrorType.PERMISSION_ERROR
        );
    }

    if (err.code === 'EXPIRED_URL') {
        throw new UploadError(
            'Presigned URL has expired',
            UploadErrorType.EXPIRED_URL,
            true
        );
    }

    throw new UploadError(
        `Upload failed: ${err.message}`,
        UploadErrorType.UNKNOWN_UPLOAD_ERROR,
        true
    );
}

Reply:
Thanks @BSalaeddin , done